dev_guide/builds/build_inputs.adoc

```diff
@@ -0,0 +1,911 @@
+[[dev-guide-build-inputs]]
+= Build Inputs
+{product-author}
+{product-version}
+:data-uri:
+:icons:
+:experimental:
+:toc: macro
+:toc-title:
+:prewrap!:
+
+toc::[]
+
+[[how-build-inputs-work]]
+== How Build Inputs Work
+
+A _build input_ provides source content for builds to operate on. There are
+several ways to provide source in {product-title}. In order of precedence:
+
+* xref:dockerfile-source[Inline Dockerfile definitions]
+* xref:image-source[Content extracted from existing images]
+* xref:source-code[Git repositories]
+* xref:binary-source[Binary inputs]
+* xref:using-secrets-during-build[Input secrets]
+* xref:using-external-artifacts[External artifacts]
+
+Different inputs can be combined into a single build. As the inline Dockerfile takes
+precedence, it can overwrite any other file named *_Dockerfile_* provided by
+another input. Binary input and Git repositories are mutually exclusive inputs.
+
+Input secrets are useful for when you do not want certain resources or
+credentials used during a build to be available in the final application image
+produced by the build. External artifacts outside of a source repository can
+also be used to pull in additional files during a build.
```

Comment:
(1) slightly.  using precedence was already problematic though...  (eg git source and binary source are mutually exclusive, so there's no applicable "precedence" there).

(2) it's not just the most common but also the most universally applicable (having an inline dockerfile w/ an s2i build doesn't usually make sense.  having a git repo always makes sense), but i think keeping the ordering symmetric is sensible.

(3) slight rewordings/augmentation:
 "Input secrets are useful for when you do not want certain resources or
credentials used during a build to be available in the final application image
produced by the build or want to consume a value that is defined in a Secret resource.  External artifacts can be used to pull in additional files that are not available as one of the other input types."